src/hash/rpo/digest.rs

```diff
@@ -47,63 +51,90 @@ impl Default for RpoDigest256 {
     }
 }
 
-impl Serializable for RpoDigest256 {
-    fn write_into<W: ByteWriter>(&self, target: &mut W) {
-        target.write_u8_slice(&self.as_bytes());
+impl Ord for RpoDigest256 {
+    fn cmp(&self, other: &Self) -> Ordering {
+        // compare the inner u64 of both elements.
+        //
+        // it will iterate the elements and will return the first computation different than
+        // `Equal`. Otherwise, the ordering is equal.
+        self.0
+            .iter()
+            .map(Felt::inner)
+            .zip(other.0.iter().map(Felt::inner))
+            .fold(Ordering::Equal, |ord, (a, b)| match ord {
+                Ordering::Equal => a.cmp(&b),
+                _ => ord,
+            })
```

Comment:
This section assumes the rpo digest is implemented as big-endian. Either that should be true, or the order of the 256 bits number shouldn't matter

Reply:
The ordering doesn't really matter we should be thinking about this as just a sequence of bytes rather than a 256-bit number. We need this primarily to using `Digest` with `BTreeMap`'s - so, as long as ordering is consistent, it should be fine.

One other issue relevant here: I think the use of `Felt::inner()` method might cause issues here. Specifically, `Felt::inner()` returns the inner representation of the element which may or may not be in canonical form. Specifically, I think for the implementation we are using, the same field element may have redundant representations (@Al-Kindi-0, correct me if I'm wrong).

A safe way would be to use `Felt::as_int()`. Though, it would be much less performant as it converts elements from Montgomery from to regular integer form.

Reply:
> If I remember correctly, the output of `Felt::new` is in the range [0,p−1] and I think all arithmetic operations produce results in the same range.

Let's double check this as this [comment](https://github.com/novifinancial/winterfell/blob/main/math/src/field/f64/mod.rs#L53) in Winterfell says otherwise. At the very least, we should update this comment.

Reply:
Thank you for catching that, it is definitely a copy-paste error from the previous implementation of f62. I have double checked both the implementation and the paper, and it is indeed the case that the values are always in $[0,p-1]$.

Reply:
If we perform the montgomery reduction there, we end up paying a bit more computation for this ord operation. Provided we are always canonical (i.e. after square, sub, add, etc), then it should be much cheaper to go with the inner representation.

We can for now perform the reduction every time, and leave an open issue to evaluate the performance gain + safety of going to the inner representation in the future.

Alternatively, I can test the full correctness of this approach now

Reply:
I think given that the elements are always canonical we can just use `Felt::inner`.

Reply:
> I have double checked both the implementation and the paper, and it is indeed the case that the values are always in [0,p−1].

Could you make a small PR against Winterfell to correct this?

Reply:
> I think given that the elements are always canonical we can just use `Felt::inner`.

Yep, that's fine then. But let's make a note in the code to state this explicitly.

Reply:
Comment updated!

Reply:
@bobbinth Sure! I will make a PR against Winterfell to correct that, it is on my TODO list.